solr/solrj/src/java/org/apache/solr/client/solrj/impl/BaseCloudSolrClient.java

```diff
@@ -862,15 +1015,39 @@ public RouteException(ErrorCode errorCode, NamedList<Throwable> throwables, Map<
     }
     List<String> inputCollections =
         collection == null ? Collections.emptyList() : StrUtils.splitSmart(collection, ",", true);
-    return requestWithRetryOnStaleState(request, 0, inputCollections);
+
+    CompletableFuture<NamedList<Object>> apiFuture = new CompletableFuture<>();
+    final AtomicBoolean cancelled = new AtomicBoolean(false);
+    final AtomicReference<CompletableFuture<NamedList<Object>>> currentFuture = new AtomicReference<>();
+    apiFuture.exceptionally((error) -> {
+      if (apiFuture.isCancelled()) {
+        synchronized (cancelled) {
```

Comment:
This code is similar to the LB class where I provided feedback.  The same feedback may apply here -- I don't see the point of synchronization on cancelled or why cancelled exists if we have a CF which can hold that cancelled state.

Reply:
I'll at-mention @CaoManhDat again whom may know as I think he authored this mechanism.

Reply:
I will provide feedback on the above LBSolrClient.